### PR TITLE
fix: Ensure the MQTT connection is properly closed

### PR DESCRIPTION
**Pull request by didier-wenzek**
## Proposed changes

Ensure the MQTT connection is properly closed

- by `tedge mqtt pub <topic> <message>`
- by `tedge mqtt sub <topic-filter>` when interrupted.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue

https://github.com/thin-edge/thin-edge.io/issues/2922

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/tedge/src/cli/mqtt/publish.rs [hunk 1]
```diff
@@ -135,6 +135,13 @@ fn publish(cmd: &MqttPublishCommand) -> Result<(), MqttError> {
     }
 
     client.disconnect()?;
+    for event in connection.iter() {
```
Comment (albinsuresh) [inline crates/core/tedge/src/cli/mqtt/publish.rs:138]: Initially I found the placement of this loop here, after the disconnect call, very confusing. But, I'm assuming that the disconnection request is also not processed until we explicitly poll the event loop as per the rumqtt design. Never expected this for a disconnect call as I thought that the `Outgoing::Disconnect` event is received by the event loop **after** the disconnect message is already sent to the broker, just as an ack to the client which doesn't need to be actively polled. But I guess they went for consistency in the processing model for disconnect as well.

File: crates/core/tedge/src/cli/mqtt/subscribe.rs [hunk 2]
```diff
@@ -108,6 +110,10 @@ fn subscribe(cmd: &MqttSubscribeCommand) -> Result<(), MqttError> {
                 eprintln!("INFO: Disconnected");
                 break;
             }
+            Ok(Event::Outgoing(Outgoing::Disconnect)) => {
+                eprintln!("INFO: Interrupted");
```
Comment (albinsuresh) [inline crates/core/tedge/src/cli/mqtt/subscribe.rs:114]:
```suggestion
                eprintln!("ERROR: Interrupted");
```

File: crates/core/tedge/src/cli/mqtt/subscribe.rs [hunk 3]
```diff
@@ -82,6 +83,7 @@ fn subscribe(cmd: &MqttSubscribeCommand) -> Result<(), MqttError> {
     }
 
     let (mut client, mut connection) = Client::new(options, DEFAULT_QUEUE_CAPACITY);
+    disconnect_if_interrupted(client.clone());
```
Comment (albinsuresh) [inline crates/core/tedge/src/cli/mqtt/subscribe.rs:86]: Why not do this for the publisher as well, to handle cases like interrupting the publisher that's stuck in the connect phase itself?

**Review by albinsuresh — COMMENTED** [review]
LGTM. Happy to approve once the query about not having a signal handler for the publisher is answered.

File: crates/core/tedge/src/cli/mqtt/subscribe.rs (hunk 2 above)
Comment (reubenmiller) [reply crates/core/tedge/src/cli/mqtt/subscribe.rs:114]: I wouldn't class this log message as an ERROR as it is normal behaviour for the user to want to stop the subscriber and the expected way is to use Ctrl-c

Comment (didier-wenzek) [reply crates/core/tedge/src/cli/mqtt/subscribe.rs:114]: I even wonder to get rid of this INFO message.

I tried to be consistent with the messages already emitted by this command (`Connected`, `Disconnect`, and now `Interrupted`). But are these really useful?

File: crates/core/tedge/src/cli/mqtt/publish.rs (hunk 1 above)
Comment (didier-wenzek) [reply crates/core/tedge/src/cli/mqtt/publish.rs:138]: This was the root cause of the issue. Triggering a `disconnect()` is not enough.

However, the point of polling the rumqtt event loop for an `Outgoing::Disconnect` event  is:
- not to perform the disconnection (this is done by rumqtt in a background thread)
- but to be sure the disconnection actually happened before the `tedge mqtt pub` command dies.

File: crates/core/tedge/src/cli/mqtt/subscribe.rs (hunk 3 above)
Comment (didier-wenzek) [reply crates/core/tedge/src/cli/mqtt/subscribe.rs:86]: Why not. A publisher will unlikely be stuck (`tedge mqtt pub` disconnects on error and awaits nothing more than an acknowledgement) but the connection could be not responsive.

=> I will add it for consistency

File: crates/core/tedge/src/cli/mqtt/publish.rs (hunk 1 above)
Comment (albinsuresh) [reply crates/core/tedge/src/cli/mqtt/publish.rs:138]: > but to be sure the disconnection actually happened before the tedge mqtt pub command dies.

In that case, should we wait for the `Event::Incoming(Incoming::Disconnect(_))` ack from the broker as well? To make sure that the broker received it as well?

Comment (didier-wenzek) [reply crates/core/tedge/src/cli/mqtt/publish.rs:138]: > In that case, should we wait for the `Event::Incoming(Incoming::Disconnect(_))` ack from the broker as well? To make sure that the broker received it as well?

No, this event is sent when the broker itself triggers the disconnection.

File: crates/core/tedge/src/cli/mqtt/subscribe.rs (hunk 3 above)
Comment (didier-wenzek) [reply crates/core/tedge/src/cli/mqtt/subscribe.rs]: Added by https://github.com/thin-edge/thin-edge.io/pull/2936/commits/531411e2ba94302723dc084b31fbc9b3cfbac523

File: crates/core/tedge/src/cli/mqtt/subscribe.rs (hunk 2 above)
Comment (rina23q) [reply crates/core/tedge/src/cli/mqtt/subscribe.rs:114]: > I even wonder to get rid of this INFO message.

I vote +1 for it.

Comment (reubenmiller) [reply crates/core/tedge/src/cli/mqtt/subscribe.rs:114]: Yeah, let's get rid of the log message. I just checked what `mosquitto_sub` does, and it does not print out a log message about being killed.

Comment (didier-wenzek) [reply crates/core/tedge/src/cli/mqtt/subscribe.rs]: Removed by https://github.com/thin-edge/thin-edge.io/pull/2936/commits/53d65369d5544b0e320a978b80863cd6a2d9b8dc

**Review by albinsuresh — APPROVED** [review]
LGTM.

**Review by reubenmiller — APPROVED** [review]
Functions nicely now, and clean disconnects for both `tedge mqtt pub` and `tedge mqtt sub` commands.

**Pull request merged**